### PR TITLE
Add SnpEff to StaPH-B/docker-builds

### PR DESCRIPTION
**Pull request by tives82**
I would like to contribute this Docker image of SnpEff. SnpEff is a variant annotation and functional effect prediction toolbox. It annotates and predicts the effects of variants on genes and proteins. This software also contains SnpSift which is a tool to filter large genomic datasets in order to find the most significant variants. 

FungiNet plans on using SnpEff to identify mutations in C. auris. They are currently working on building a C. auris genome database for SnpEff that will identify FKS1 mutations. SnpEff uses VCF files as input (a mycosnp_nf output).

- [ X] This comment contains a description of what is in the pull request.

- [ X] Build your own docker image using a Dockerfile
  - [X ] Directory structure should be name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. spades/3.12.0/Dockerfile)
  - [ X] Includes the recommended LABELS
- [X ] Edit main README.md
- [X ] Edit Program_Licenses.md
- [X ] Create a simple container-specific README.md in the same directory as the Dockerfile (i.e. spades/3.12.0/README.md)
- [ X] Write a GitHub actions workflow
  - [X ] Should be located in .github/workflows/ and named test-<tool>.yml (i.e. .github/workflows/test-spades.yml)
  - [X ] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
  - [ X] Have successfully run the workflow "Test <program name> image" in your forked repository

### CODE REVIEW

**Comment by kapsakcj** [comment]
Hi Tom, thanks for the PR! Appreciate the info about FungiNet - I've been wondering how resistance mutations are identified in C. auris so this is super helpful.

I'm going to make a few comments on your code and suggest some changes.

It looks like the GH actions workflow for testing did not get run, so I think the YML will need some tweaking.

File: .github/workflows/test-snpeff.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# Some explanations come from: https://github.com/actions/starter-workflows/blob/main/automation/manual.yml
+
+name: Test snpeff image
+
+# Controls when the action will run. Workflow runs when manually triggered using the UI or when you submit your pull request
+on:
+  workflow_dispatch:
+  pull_request:
+    paths:
+      - "<snpeff/5.1/Dockerfile"  # Dockerfile path, e.g. 'htslib/1.14/Dockerfile' so that only your image is tested
```
Comment (kapsakcj) [inline .github/workflows/test-snpeff.yml:12]: I think the `<` needs to be removed from this line

File: .github/workflows/test-snpeff.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
+
+  # This job calls a workflow to build the image to the 'test' stage
+  build-to-test:
+    uses: StaPH-B/docker-builds/.github/workflows/build-to-test.yml@master
```
Comment (kapsakcj) [inline .github/workflows/test-snpeff.yml:19]: If you change this line to: `./.github/workflows/build-to-test.yml` then you'll be able to run the testing workflow on your forked repo.

The way it is currently set is restricted to the StaPH-B code repository

File: snpeff/5.1/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+RUN wget http://sourceforge.net/projects/snpeff/files/snpEff_latest_core.zip && \
+  unzip snpEff_latest_core.zip && \
+  rm -rf snpEff_latest_core.zip && \
+  chmod +x /snpEff/snpEff.jar && \
+  echo "#!/bin/bash" >> /snpEff/snpeff && \
+  chmod +x /snpEff/SnpSift.jar && \
+  echo "#!/bin/bash" >> /snpEff/snpsift && \
+  echo "exec java -jar /snpEff/snpEff.jar """"$""@"""" " >> /snpEff/snpeff && \
+  chmod +x /snpEff/snpeff && \
+  echo "exec java -jar /snpEff/SnpSift.jar """"$""@"""" " >> /snpEff/snpsift && \
```
Comment (kapsakcj) [inline snpeff/5.1/Dockerfile:35]: nice hack on the java executables, I love it 😄

**Comment by kapsakcj** [comment]
I've created the blank repos on dockerhub and quay in preparation for this docker image

https://hub.docker.com/r/staphb/snpeff
https://quay.io/repository/staphb/snpeff

File: snpeff/5.1/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+# Install open using apt 
+RUN apt-get update && apt-get install -y software-properties-common &&\
+  add-apt-repository -y ppa:openjdk-r/ppa &&\
+  apt-get update && apt-get install -y \
+  openjdk-11-jre \
+  wget \
+  unzip
+
+# Use wget to install SnpEff and SnpSift
+RUN wget http://sourceforge.net/projects/snpeff/files/snpEff_latest_core.zip && \
```
[the remark this answers is not in the document]
Comment (kapsakcj) [reply snpeff/5.1/Dockerfile:26]: Related: https://github.com/pcingola/SnpEff/issues/392

It may not be possible at the moment, but perhaps I am missing something

[the remark this answers is not in the document]
Comment (kapsakcj) [reply snpeff/5.1/Dockerfile:26]: additionally I noticed that the jar files downloaded via the link in the dockerfile downloads an older version of snpEff (or at least the `-version` flag prints an older version

```
$ wget http://sourceforge.net/projects/snpeff/files/snpEff_latest_core.zip

$ unzip snpEff_latest_core.zip

$ cd snpEff/

$ java -jar snpEff.jar -version
SnpEff  4.3t    2017-11-24

```

File: snpeff/5.1/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+LABEL description.SnpEff="Genetic variant annotation and effect prediction toolbox."
+LABEL description.SnpSift="Used after SnpEff annotation to filter and manipulate annotated files."
+LABEL website="https://pcingola.github.io/SnpEff"
+LABEL license="https://github.com/pcingola/SnpEff/blob/master/LICENSE.md"
+LABEL maintainer="Tom Iverson"
+LABEL maintainer.email="tiverson@utah.gov"
+
+# Install open using apt 
+RUN apt-get update && apt-get install -y software-properties-common &&\
+  add-apt-repository -y ppa:openjdk-r/ppa &&\
```
Comment (kapsakcj) [inline snpeff/5.1/Dockerfile:19]: I'm not 100% sure, but I don't think you need this line for `add-apt-repository -y ppa:openjdk-r/ppa &&\`

openjdk-11-jre should be readily available via `apt` for ubuntu:focal https://packages.ubuntu.com/focal/openjdk-11-jre

File: snpeff/5.1/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (tives82) [reply snpeff/5.1/Dockerfile:26]: Yeah, I've looked into this quite a bit because I initially wanted to download a tagged v5.1 but I can't find a v5.1 anywhere, only the "snpeff_latest_core" which apparently is v4.3t.

In fact, [https://pcingola.github.io/SnpEff/](https://pcingola.github.io/SnpEff/) is the only place I find the 5.1 version stated. Like you said, the jar files -version flag prints v4.3t and the snpEff.config states to check [http://snpeff.sourceforge.net/versions.txt](http://snpeff.sourceforge.net/versions.txt) for the latest version.
```SNPEFF	4.3	2019-01-26	http://sourceforge.net/projects/snpeff/files/snpEff_latest_core.zip ```

I was thinking that "snpEff_latest_core" is tagged to always be the latest version. Hopefully you get a reply to [pcingola/SnpEff#392](https://github.com/pcingola/SnpEff/issues/392). I've asked Pablo Cingolani to connect on LinkedIn so I can message him and ask.

File: snpeff/5.1/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+# Set environment variable $PATH
+ENV PATH="${PATH}:/snpEff" 
+
+# Set working directory
+RUN mkdir /data
+WORKDIR /data
+
+# Run test of annotation
+From app as test
```
[the remark this answers is not in the document]
Comment (tives82) [reply snpeff/5.1/Dockerfile]: Nice catch. Thanks!

File: snpeff/5.1/Dockerfile (hunk 5 above)
Comment (tives82) [reply snpeff/5.1/Dockerfile]: You're right. It's available and installs fine. Thanks!

File: README.md [hunk 7]
```diff
@@ -174,3 +175,4 @@ Each Dockerfile lists the author(s)/maintainer(s) as a metadata `LABEL`, but the
   * [@SarahNadeau](https://github.com/SarahNadeau)
   * [@DOH-HNH0303](https://github.com/DOH-HNH0303)
   * [@jwarnn](https://github.com/jwarnn)
+  * [@tiverson](https://github.com/tives82)
```
Comment (erinyoung) [inline README.md:178]: Yay! I love it when we have new contributors!

File: snpeff/5.1/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (erinyoung) [reply snpeff/5.1/Dockerfile:25]: I think it might be easier to download from github and install from source. You would have to install ANT and MAVEN during the apt-get phase.
```
snpeff_ver=5.1d

# get the tar file for the desired version:
wget https://github.com/pcingola/SnpEff/archive/refs/tags/v${snpeff_ver}.tar.gz && \
# unzip the tar file
tar -xvf ${snpeff_ver}.tar.gz && \
# remove the tar file (since it's not needed anymore)
rm ${snpeff_ver}.tar.gz 
```
And then use maven (mvn) to install the other dependencies as described in https://pcingola.github.io/SnpEff/download/#building-from-the-source

File: snpeff/5.1/galaxy.sh [hunk 8]
```diff
@@ -0,0 +1,18 @@
+#!/bin/sh
```
Comment (erinyoung) [inline snpeff/5.1/galaxy.sh:1]: Why is this file needed?

File: snpeff/5.1/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  rm v${SNPSIFT_VER}.tar.gz
+
+# Build libraries from the source code using Maven.
+RUN cd SnpEff-${SNPEFF_VER}/lib && \
+  mvn install:install-file -Dfile=antlr-4.5.1-complete.jar -DgroupId=org.antlr -DartifactId=antlr -Dversion=4.5.1 -Dpackaging=jar && \
+  mvn install:install-file -Dfile=biojava3-core-3.0.7.jar -DgroupId=org.biojava -DartifactId=biojava3-core -Dversion=3.0.7 -Dpackaging=jar && \
+  mvn install:install-file -Dfile=biojava3-structure-3.0.7.jar -DgroupId=org.biojava -DartifactId=biojava3-structure -Dversion=3.0.7 -Dpackaging=jar
+
+# Once libraries are installed, use make.sh to build the code.
+COPY make.sh .
```
Comment (erinyoung) [inline snpeff/5.1/Dockerfile:46]: Instead of using a make.sh file that you've included, could you use the make.sh that's part of their github repo instead of this one? It'd make things easier for the next snpeff update.

Although... why are you using make.sh when you're installing the dependencies in lines 41-43.

File: snpeff/5.1/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+LABEL maintainer.email="tiverson@utah.gov"
+
+# Install open using apt 
+RUN apt-get update && apt-get install -y software-properties-common &&\
+  apt-get update && apt-get install -y \
+  openjdk-11-jre \
+  ant \
+  maven \
+  wget \
+  unzip
```
Comment (erinyoung) [inline snpeff/5.1/Dockerfile:29]: Could you add 
```
&& \
 apt-get autoclean && rm -rf /var/lib/apt/lists/*
 ```
 
 to the end of the apt-get installation step? This helps reduce the size of the container.

File: .github/workflows/test-snpeff.yml [hunk 11]
```diff
@@ -0,0 +1,23 @@
+# This caller workflow builds an image to the "test" stage.
```
Comment (erinyoung) [inline .github/workflows/test-snpeff.yml:1]: We don't need this file anymore. It works! It's great! Could you delete it?

File: snpeff/5.1/Dockerfile (hunk 9 above)
Comment (tives82) [reply snpeff/5.1/Dockerfile]: I've added the code to build the snpeff/snpsift jar executables instead of copying the make.sh file, as well as removed the galaxy.sh and test-snpeff.yml files. Hopefully, it should be good to go now.

File: snpeff/5.1/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  rm v${SNPEFF_VER}.tar.gz && \ 
+  wget https://github.com/pcingola/SnpSift/archive/refs/tags/v${SNPSIFT_VER}.tar.gz && \
+  tar -xvf v${SNPSIFT_VER}.tar.gz && \
+  rm v${SNPSIFT_VER}.tar.gz
+
+# Install dependencies and snpEff/SnpSift executables from the source code using Maven.
+RUN cd SnpEff-${SNPEFF_VER}/lib && \
+  mvn install:install-file -Dfile=antlr-4.5.1-complete.jar -DgroupId=org.antlr -DartifactId=antlr -Dversion=4.5.1 -Dpackaging=jar && \
+  mvn install:install-file -Dfile=biojava3-core-3.0.7.jar -DgroupId=org.biojava -DartifactId=biojava3-core -Dversion=3.0.7 -Dpackaging=jar && \
+  mvn install:install-file -Dfile=biojava3-structure-3.0.7.jar -DgroupId=org.biojava -DartifactId=biojava3-structure -Dversion=3.0.7 -Dpackaging=jar
```
Comment (erinyoung) [inline snpeff/5.1/Dockerfile:45]: I think line 45 is missing a '\' at the end of it.

**Review by erinyoung — APPROVED** [review]
Phew!

**Comment by erinyoung** [comment]
Passed all tests and looks good to me. I'm merging this and deploying to dockerhub. It should be there _soon_
